man/yac2ncid.1
.\" %W% %G%
.TH YAC2NCID 1
.SH NAME
.B yac2ncid
- YAC server to NCID server gateway
.SH SYNOPSIS
.B yac2ncid
[-v]
.SH DESCRIPTION
.B YAC2NCID
acts as a YAC client, listens to the CID information from a YAC server
and sends it to the NCID server.  The server then sends the CID
information to the  NCID clients.
.PP
It is important to understand that to run YAC server with NCID
clients you must run YAC2NCID gateway as well as NCID server.
.PP
The configuration file for \fIyac2ncid\fR is \fI/etc/ncid/yac2ncid.conf\fR.
It may be located in same directory as the script, or in the
\fIetc\fR directory, or in NCID default configuration directory.
Read the comments in the config file for more details.
.PP
.B YAC2NCID
can run on any system, but normally it is run on same box as the NCID server.
If it is not run on the same box as the NCID server, you must configure
the server IP address in the configuration file.
.PP
Configure the YAC server with IP address of the box running
.BR YAC2NCID .
.PP
.B YAC2NCID
relies on netcat binary (nc). As there are multiple implementations of
this binary in existence, the script will try to auto detect the options
to use. If script is unable to properly detect the version of nc, set
NETCAT_TYPE to CUSTOM in the config file and set custom parameters to
match your version.
.PP
There is also a client output module called \fIncid-yac\fR.  It sends
the NCID CID information to YAC clients.
.PD
.SH "OPTIONS"
.PD 0
.TP 10
.B -v
Verbose mode.
.PD
.PD
.SH REQUIRED PROGRAMS
netcat: /usr/bin/nc
.SH FILES
/etc/ncid/yac2ncid.conf
.SH SEE ALSO
ncid2ncid.1, yac2ncid.1, ncid.1, ncidrotate.1, ncidtools.1
yac2ncid.conf.5,
ncidtools.7,
ncidd.8, sip2ncid.8


doc/CHANGES
ncid - Network CallerID

Version 0.87
    - README: updated
    - Makefile: added test: $(DOCDIR) does not exist, create it
    - All GPL license headers in files were updated to GPL 3
    - client/ncid.sh: logClock{} changed {^(CID|HUP|OUT)} to {^(?:CID|HUP|OUT)}
                      added "Line Labels" in help menu and labInfo {}
                      checkTypes() added new types and changed type numbers
                      getCID(): added new types and changed if statements
                                for new Type numbers
    - debian/Makefile: updated
    - debian/README.Debian: updated
    - debian/*.init: removed "S" from default STOP
    - debian/wc2ncid.init: removed configuration stuff, changed sbin to sbin
    - debian/yac2ncid.init: removed configuration stuff, changed sbin to sbin
    - debian/rn2ncid.init: new
    - doc/INSTALL-Ubuntu: updated
    - doc/INSTALL-Raspbian: updated
    - doc/LICENSE: updated to GPL 3
    - doc/Setup-rn2ncid: new
    - doc/Verbose-rn2ncid: new
    - Fedora/Makefile: updated
    - Fedora/README.Fedora: updated
    - Fedora/rn2ncid.init: new
    - Fedora/rn2ncid.service: new
    - FreeBSD/Makefile: updated
    - FreeBSD/README.FreeBSD: updated
    - FreeBSD/rn2ncid.rc: new
    - gateway/Makefile: added $(EXTRA_CFLAGS) to CFLAGS definition
    - gateway/README.Gateways: renamed to README.gateways and updated
    - gateway/*.[ch]: updated license in headers
    - gateway/*.pl: updated license in comments
    - gateway/sip2ncid.h: added UNAUTHORIZED define
    - gateway/sip2ncid.c: processPackets() added UNAUTHORIZED section
    - gateway/wc2ncid.conf: - improved comment wording
                            - changed nciddelay to delay
    - gateway/wc2ncid.pl: - moved test mode just after Command Line Processing
                          - moved logfile and config file messages to
                            after start message
                          - added cleanup(), called by errorExit() & sigHandle()
                          - added delay option to pod documentation wording,
                            improved some pod wording
                          - changed $nciddelay to $delay, added $Delay,
                            added --delay|-d option
    - gateway/rn2ncid.pl: new
    - gateway/rn2ncid.config: new
    - Mac/Makefile: updated
    - Mac/README.Mac: updated
    - Mac/net.sourceforge.rn2ncid.plist: new
    - man/Makefile: updated
    - man/ncid-*.1: updated
    - man/ncid.1: updated
    - man/ncid.conf.5: updated
    - man/ncidd.8: updated
    - man/yac2ncid.1: updated
    - man/rn2ncid.conf.5: new
    - modules/ncid-*.sh: updated
    - modules/ncid-*.conf: updated
    - modules/ncid-notify.sh: changed line one from /bin/sh to /bin/bash
                              minor changes to verbose output and comments
    - modules/ncid-speak.sh: added "AreaCodeLength" code to speak the area
                             code instead or "NO NAME" or entire number
    - modules/ncid-speak.conf-in: added "AreaCodeLength" and comments
    - server/ncidd.c: writeLog() - moved "logMsg(LEVEL3, msgbuf);" to start
                      formatCID() - moved cidnoname test inside
                                    if (!(cid.status & CIDNMBR))
                                 - added PID:
                      added BLKLINE, NOTLINE, PIDLINE to *lineTags[]
                      doPoll: - added code for a NOT: message
    - server/ncidd.h: added defines for PID, PIDLINE, NOTLINE, CALLPID
    - test/README.test: updated
    - test/*: renamed test programs and test data
    - test/test-rn2ncid: new
    - test/test-rn2ncid.data: new
    - TiVo/ncid-tivo.sh:  if no line label, display $CIDNAME/$CIDNMBR
                          else display "$CIDNAME $CIDNMBR/$CIDTYPE $CIDLINE"

Version 0.86.1
    - debian/changelog: updated
    - debian/control: for package ncid-client added "curl" to Depends

Version 0.86
    - README: updated
              added redhat: and redhat-install:
    - attic/*: new
    - client/ncid.sh: added more doVerbose statements and levels
                      changed 300* doVerbose message to any one of four
                      added line count to "End of call log" message
                      removed the --raw | -R option
                      fixed and improved call on a ring number
                      fonts used depend on platform and OS
                      allow resizing of window
                      allow font size changes
                      time format in history window can be either 12 and 24
                      new menu called Preferences for Fonts and time format
                      menu item to save window size
                      menu item to save window size and location
                      added the "-h | --help" option
                      added the "-v | --version" option
                      added the "-w | --wakeup" option
                      getCID{}: added calls to ncid-wakeup module
                      changed "-V" to be the "-v | --verbose" option
                      makeWindow{}: improved widget and menu code
                      added saveSize{} write_rc_file{} logFont{} changeFont{}
                      getField{}: changed regex to include "*" and "-"
    - client/ncid.conf-in: removed MsgFlag variable and comments
                           removed Raw variable and comments
                           added comments to CallOnRing section
                           added comments to Ring Options section
                           new "WakeUp" section
                           new "WrapLines" section
    - debian/Makefile: updated
    - debian/wc2ncid.init: new
    - debian/ncidsip.init: removed
    - doc/CONTRIBUTORS: updated
    - doc/NCID-SDK.odt: updated
    - doc/Setup-ncidsip: removed
    - doc/Setup-sip2ncid: updated
    - doc/Setup-wc2ncid: new
    - doc/INSTALL-Redhat: new
    - doc/INSTALL*: updated
    - doc/INSTALL-RPi: new
    - Fedora/Makefile: updated

cidgate/yac2ncid.sh
#!/bin/bash
##
## YAC2NCID - YAC to NCID gateway service
##
## NOTE: This script requires netcat binary (nc) to operate
##       The bash location in the first line must match the if test below
##
## Copyright (c) 2007, 2008
## by Michael Lasevich(hichhiker)
## Version: 1.0 (6/14/2007)
## Version: 1.0a (6/26/2007)
##
## Version: 1.0.[123] (5/29/2008) by John L. Chmielewski
## Copyright (c) 2010, 2012
## Last changed on Thu Jul 5, 2012 by jlc
##
VERSION="(NCID) XxXxX"

ME="$(basename $0)"
ME="${ME%.sh}"
MYDIR="$(cd $(dirname $0) && pwd -P)"

## Workaround for broken trap on TiVo
## checking if date is 199?: broken bash
## could also test for TiVo: /tvbin/tivoapp
if /bin/bash --version | grep 199 > /dev/null
then
    # must use bash with broken trap
	PIDFILE=/dev/null
else
    # trap should be OK
    PIDFILE=/var/run/yac2ncid.pid

    dienice(){
        if test -f ${PIDFILE}
        then
            local line pid
            echo >> ${PIDFILE}
            read line < "${PIDFILE}"
            for pid in $line ; do
                [ -z "${pid//[0-9]/}" -a -d "/proc/$pid" ] &&
                [ $pid != $$ ] &&
                kill $pid
            done
		    rm -f ${PIDFILE}
        fi
        exit
    }

	trap dienice INT HUP QUIT TERM EXIT
fi

set -e

CONFFILE=""

ConfigDir=/usr/local/etc/ncid

VERBOSE=

YACPORT=10629
NCIDSERVER=localhost:3333
LINE="YAC"
NETCAT=

NETCAT_SEND_OPTS=""
NETCAT_LISTEN_OPTS="-p"

NETCAT_BUSYBOX_SEND_OPTS=""
NETCAT_BUSYBOX_LISTEN_OPTS="-p"

NETCAT_HOBBIT_SEND_OPTS="-w1"
NETCAT_HOBBIT_LISTEN_OPTS="-p"

NETCAT_JACKSON_SEND_OPTS="-w1"
NETCAT_JACKSON_LISTEN_OPTS=""

NETCAT_BSD_SEND_OPTS="-w1"
NETCAT_BSD_LISTEN_OPTS=""

NETCAT_CUSTOM_SEND_OPTS="-w1"
NETCAT_CUSTOM_LISTEN_OPTS="-p"

NCIDHOST="localhost"
NCIDPORT="3333"

##
## isblank(string) - returns true if string is blank
##

isblank(){
	test "z${*}" == "z"
}


##
## msg(string) - sends message
##

msg(){
	isblank $VERBOSE ||\
		echo ${ME}: $*
	true
}



##
## findfile(file, path) - find "file" in "path"
##
findfile(){
	local filename="${1}"
	local path="${2}"
	local O_IFS="${IFS}"
	local N_IFS=":"
	local fullfile=""

	IFS="${N_IFS}"
	for location in ${path:-.}
	do
		IFS="${O_IFS}"
		local fullfile="${location}/${filename}"
		test -f "${fullfile}" &&\
			echo ${fullfile} &&\
			break
		IFS="${N_IFS}"
	done
	IFS="${O_IFS}"
}

## Process a message
procMSG(){
	local nciddate=$(date +"%m%d%H%M")
	local ncidname=""
	local ncidnnumber=""

	local NCIDMSG="${*}"	

	if ! isblank $(echo "${*}" | grep '^@CALL')
	then
		ncidname=$(echo ${NCIDMSG} | cut -c6- | cut -d'~' -f1);
		ncidnumber=$(echo ${NCIDMSG} | cut -s -d'~' -f2-);
		NCIDMSG="$(printf 'CALL: ###DATE%s...LINE%s...NMBR%s...NAME%s+++' "${nciddate:-01010000}" "${LINE:--}" "${ncidnumber:-unlisted}" "${ncidname:-noname}")"
    elif ! isblank "${*}"
    then
        NCIDMSG="MSG: ${NCIDMSG}"
	fi	
  	echo ${NCIDMSG} | send2ncid
}



## send stdin to ncidd server
send2ncid(){
	${NETCAT:-nc} ${NETCAT_SEND_OPTS} ${NCIDHOST:-localhost} ${NCIDPORT:-3333} >/dev/null
}


## Check command line options

test "z${1}" == "z-v" &&\
	VERBOSE="yes" &&\
	msg Enabling verbose mode due to command line option;\
    msg Version ${VERSION}


CONFFILE="$(findfile ${ME%.sh}.conf "${MYDIR}:${MYDIR}/../etc:${ConfigDir}:~:/etc")"

isblank ${CONFFILE} &&\
    msg ${ME%.sh}.conf not found ||\
    {
	    msg Using ${CONFFILE}
        source ${CONFFILE}
    }


## Parse NCIDSERVER
NCIDHOST="$(echo ${NCIDSERVER:-localhost} | cut -d: -f1)"
NCIDPORT="$(echo ${NCIDSERVER:-localhost} | cut -s -d: -f2)"
msg Using NCID host: ${NCIDHOST}
msg Using NCID port: ${NCIDPORT}
msg Using YAC port: ${YACPORT}


## Make sure netcat is available
isblank ${NETCAT} &&\
	NETCAT=$(which nc 2>/dev/null || echo nc)

! test -x ${NETCAT:=nc}  &&\
	echo Unable to find netcat binary \(${NETCAT}\) &&\
	exit -1

if isblank ${NETCAT_TYPE}
then
	msg Auto-detecting the netcat type
	if test "z$(${NETCAT:-nc} -h 2>&1  | head -1 |  cut -d' ' -f1)" == 'z[v1.10]'
	then
		msg Detected \*Hobbit\* version of netcat 
		NETCAT_TYPE="HOBBIT"
	elif ! isblank "$(${NETCAT:-nc} -h 2>&1 | grep BusyBox)"
	then	
		msg Detected BusyBox version of netcat 
		NETCAT_TYPE="BUSYBOX"
	elif ! isblank "$(${NETCAT:-nc} -h 2>&1 | \
           egrep '46CDdhklnrStUuvz|46DdhklnrStUuvzC')"
	then
		msg Detected \"Eric Jackson\" version of netcat
		NETCAT_TYPE="JACKSON"
	elif ! isblank "$(${NETCAT:-nc} -h 2>&1 | grep '46DEdhklnrStUuvz')"
	then
		msg Detected \"BSD\" version of netcat
		NETCAT_TYPE="BSD"
	else
		msg Unknown version of netcat, using CUSTOM values
		NETCAT_TYPE="CUSTOM"
	fi
fi


case "${NETCAT_TYPE:-CUSTOM}" in
	"BUSYBOX"|"busybox")
		msg Using BusyBox settings...
		NETCAT_SEND_OPTS="${NETCAT_BUSYBOX_SEND_OPTS}"	
		NETCAT_LISTEN_OPTS="${NETCAT_BUSYBOX_LISTEN_OPTS}"	
		;;
	"HOBBIT"|"hobbit")
		msg Using Hobbit settings...
		NETCAT_SEND_OPTS="${NETCAT_HOBBIT_SEND_OPTS}"	
		NETCAT_LISTEN_OPTS="${NETCAT_HOBBIT_LISTEN_OPTS}"	
		;;
	"JACKSON"|"jackson")
		msg Using Jackson settings...
		NETCAT_SEND_OPTS="${NETCAT_JACKSON_SEND_OPTS}"	
		NETCAT_LISTEN_OPTS="${NETCAT_JACKSON_LISTEN_OPTS}"	
		;;
	"BSD"|"BSD")
		msg Using BSD settings...
		NETCAT_SEND_OPTS="${NETCAT_BSD_SEND_OPTS}"	
		NETCAT_LISTEN_OPTS="${NETCAT_BSD_LISTEN_OPTS}"	
		;;
	*)
		msg Using custom settings...
		NETCAT_SEND_OPTS="${NETCAT_CUSTOM_SEND_OPTS}"	
		NETCAT_LISTEN_OPTS="${NETCAT_CUSTOM_LISTEN_OPTS}"	
		;;
	
esac

msg Listening for new connections
## Main loop
while true
do
	MSG=$(bash -c "echo \"\$\$ \">>${PIDFILE:-/tmp/ncpid} && exec ${NETCAT:-nc} -l ${NETCAT_LISTEN_OPTS} $YACPORT")
	test "z" != "z${MSG}" &&\
		msg Got a message &&\
		procMSG "$MSG" &
done
